project-repo improvements (#5)
canonical/desktop-engineering: project-repo/review-process.md

# Review process

## General principles

In general direct pushes to the `main` branch are not allowed, and Pull Requests should be used. The only allowed push to `main` is to push the tag for a new release. Optionally, the direct push of an emergency fix to `main` may be allowed, but a good reason is needed for bypassing the pull request requirement.

> Even in those cases, it’s still better to go through a pull request to ensure that automated tests and checks are running, even if it’s merged without further reviews.

Each pull request should have at least one reviewer. This is our way of ensuring quality and double-checking on what we produce as a team. This applies both to the projects we are the upstream for or we are a contributor to, as well as to packaging (snap or .deb) changes.

When contributing to the projects we are not the upstream for, we of course follow upstream's methodology for code reviews. However, the guidelines below can still help ensure that any submission made by a Canonical contributor is of high quality right from the get-go.

> If a change has been done in a pair-programming context, that can be considered as already reviewed.  However, sanity checks (see the first steps) are still expected to be carried out by both parties, before merging.

Reviewer: ```suggestion
If a change has been done in a pair-programming context, that can be considered as already reviewed. However, sanity checks (see the first steps) are still expected to be carried out by both parties, before merging.
```

### Policies and guidelines

The aim of these guidances is to ease the maintainability of code and the project as a whole, in the long run. Double questioning an approach or providing suggestions should always follow this trail of thoughts: How will it make the code easier to understand and maintain, and more robust?  Can we avoid premature optimisations towards that goal?

* Always prefer to land multiple small pull requests, rather than a large pull request with multiple unrelated changes mixed in. Both your future self and the reviewer will thank you.
* As a consequence, we don’t encourage "long-lived feature branch".

> * Any new code and fixes should be accompanied with new or modified existing tests.  Only fixes for rare crashes that are exceptionally difficult to reproduce are acceptable without accompanying automated tests.  If that is the case for your particular fix, please ensure the description reflects this.

Reviewer: ```suggestion
* Any new code and fixes should be accompanied with new or modified existing tests. Only fixes for rare crashes that are exceptionally difficult to reproduce are acceptable without accompanying automated tests. If that is the case for your particular fix, please ensure the description reflects this.
```

* Any commits merged in the `main` branch need to pass the tests. This is our mainline. You can split your changes into multiple, more discrete modifications, so as to help give additional context when one needs to dig into the git history later on.
* Reviewing a change is serious business. A generic "LGTM" should be more the exception than the rule. Comments are encouraged and should be seen as positive feedback to broaden our understanding of the issue at hand. Accordingly, the commenter should strive to make their feedback as constructive as possible.

> * When there are multiple reviewers on a PR, please pay attention to the comments and answers of your peers!  This helps building shared understanding of the code. All of us could always learn something new. :-)

Reviewer: ```suggestion
* When there are multiple reviewers on a PR, please pay attention to the comments and answers of your peers! This helps building shared understanding of the code. All of us could always learn something new. :-)
```

### Merge workflow

We should use the **merge commit workflow** (not squash and merge). The history of each branch should make sense on its own, with few or no "fixup" commits.

However, for better readability of the mainline, the merge commit in the mainline should follow [conventional commits][convcommits]. This can be enforced [on a per project][gh-docs-merge-commit] basis on GitHub.

> TODO: have a document defining what to configure in every new GitHub project (webhook for Jira, merge commit + default commit template, automated merged branch deletion).

[convcommits]: https://www.conventionalcommits.org/en/v1.0.0/
[gh-docs-merge-commit]: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-merging-for-pull-requests

## Process

Don't hesitate to comment and reiterate! When you engage with a pull request, the expectation is that you remain responsive, and not let the pull request go stale.

> 1. The pull request should always be opened as a draft, to not notify the reviewers from automation. Ensure the description of changes is readable and understandable by prospective reviewers. The pull request should be linked to a Jira card, using `UDENG-XXX` in its description, so that Jira links against it.

Reviewer: ```suggestion
1. The pull request should always be opened as a draft, to avoid notitying the reviewers before the tests pass and the pull request is fully ready. Ensure the description of changes is readable and understandable by prospective reviewers. The pull request should be linked to a Jira card, using `UDENG-XXX` in its description, so that Jira links against it.
```

> 1. The pull request diff should be examined by the submitter - it may happen that a debug print statement was left in, or a non-related change was accidentally `git add`ed. Please at least do a quick visual scan to help catch such potential mistakes.

Reviewer: isn’t `mistakes` a little bit too strong in that context?

Author: Hmm, that didn't occur to me, but sure, I've changed it to `issues/typos`. Is that better?

Reviewer: sounds good to me!

1. Any changes should be appended to the commit history, ready to be squashed into potential relevant existing commits. The new changes should be kept as separate commits for the duration of the review, to ease further reviews, as most tools don’t have a good diff-of-diffs view.
1. The submitter assesses the CI test results and code coverage, and requests another review.
1. The reviewer examines the changes, and resolves any fixed comments or agrees with the arguments from the submitter.
   > It’s always the person opening a comment, **not** the submitter of the pull request, that should resolve the comments, once they are happy with the particular change.

This cycle is repeated if there are further comments or if some are still left open. Once all comments are resolved, the reviewer should approve.

### Final merge

1. The submitter will squash the optional additional commits and force push to the pull request.
1. The submitter examines one last time the diff, the test results, and the coverage.
1. The submitter is in charge of merging the changes by clicking the merge button.

> > Note: The final merge steps may be proceeded by the reviewer if we are in a hurry. However, it’s preferable to allow the submitter to do the actual merging, as an additional way of acknowledging their their work and enjoying the serotonin boost. :-)

Reviewer: their their*

* Try to not let an open pull request remain without activity for more than 3 days. Especially, try to be responsive within a day for contributions from external contributors. Stale branches should be either worked on or deleted. This helps ensure that the repository has an easy and actionable pull request review list.
* Automated pull requests for dependency updates should be looked at with scrutiny: tests passing does not necessarily imply that the pull request is good to merge. Examine at the upstream changelog and/or commit logs, and assess the impact to see if we would benefit from the changes, and generally aim to reduce technical debt. If any changes are required to make the tests pass again for such a pull request, those changes should be added to the very same branch, so that merging the pull request would not leave the tests broken.

### Merge workflow

We should use the **merge commit workflow** (not squash and merge). The history of each branch should make sense on its own, with few or no "fixup" commits.

> However, for better readability of the mainline, the merge commit in the mainline should follow [conventional commits][convcommits]. This can be enforced [on a per project][gh-docs-merge-commit] basis on GitHub.

Reviewer: I know the original text was mentioning this conventional commits. I still need to dig if this is what we really want or not (I think we need to change this as I tried to experiment with it and I am not convinced at all). do you mind adding a FIXME here?

> TODO: have a document defining what to configure in every new GitHub project (webhook for Jira, merge commit + default commit template, automated merged branch deletion).

> [convcommits]: https://www.conventionalcommits.org/en/v1.0.0/

Reviewer: Having the reference just after the section doesn’t bring much. I think it’s either we have a link section at the top of the page for easily check that, either it’s inline. Just not at the bottom of each section IMHO.

Author: Thanks. I kinda/mostly agree with your point, except instead of putting them at the top, let's put them at the very bottom. That's the convention I've seen used most elsewhere as well, and makes sense (kind of like how in an article, the 'references' section goes at the very end).  I've moved these to the bottom of the page.

I think our recommendation should be: short-ish links (e.g. to other md files in the same repo) are okay inline, but links with long URLs should go to the bottom of the page.

Reviewer: I think of them mostly as constant in code, hence the top (a little bit like in the project template, as this is the links that you are going to change anyway when creating a new repo instance). However, as long as they are all grouped, I don’t really have any strong opinion and I’m fine with at the bottom.

Agreed on the short vs link links.
